src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs

```diff
@@ -130,7 +134,11 @@ public async Task SetInstalledTemplatePackagesAsync(IReadOnlyList<TemplatePackag
 
                 try
                 {
+                    // Ignore FSW notifications received during writing changes (we'll notify synchronously)
+                    Interlocked.Increment(ref _waitingInstances);
                     _environmentSettings.Host.FileSystem.WriteObject(_globalSettingsFile, globalSettingsData);
+                    // We are ready for new notifications now
+                    Interlocked.Exchange(ref _waitingInstances, 0);
                     SettingsChanged?.Invoke();
```

Comment:
is it needed? I guess due to this line we can possibly invoke event twice: once here and one in `FileChanged`.
Should we raise FileChanged here instead?

Reply:
Yes - there is an opportunity for a race. It's just best effort to improve perf of this scenario, but it's not creating consistency problems if event is called multiple times in parallel (listener needs to acquire the lock first)

Reply:
Added improved version with deregistering FSW